<a href="https://colab.research.google.com/github/smartwatch11/ML/blob/main/HSE_Exam_EgorRybin.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
hash('evrybin@edu.hse.ru') % 3

1

My machine learning task is predicting class of Mushroom (edible or poisonous)

In [ ]:
# Importing the required library for data processing, works with CSV file
import pandas as pd

# Reading the dataset
df = pd.read_csv('mushrooms.csv', error_bad_lines=False)

In [ ]:
df.head() #See what the data is

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0     p         x           s         n       t    p               f   
1     e         x           s         y       t    a               f   
2     e         b           s         w       t    l               f   
3     p         x           y         w       t    p               f   
4     e         x           s         g       f    n               f   

  gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0            c         n          k  ...                        s   
1            c         b          k  ...                        s   
2            c         b          n  ...                        s   
3            c         n          n  ...                        s   
4            w         b          k  ...                        s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

Since we are dealing with categorical values, it is obvious that it will be necessary to create a boolean column for each property (1-if there is a feature, 0-if there is no given feature)

In [ ]:
df.describe() #One more summary information

class cap-shape cap-surface cap-color bruises  odor gill-attachment  \
count   8124      8124        8124      8124    8124  8124            8124   
unique     2         6           4        10       2     9               2   
top        e         x           y         n       f     n               f   
freq    4208      3656        3244      2284    4748  3528            7914   

       gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
count          8124      8124       8124  ...                     8124   
unique            2         2         12  ...                        4   
top               c         b          b  ...                        s   
freq           6812      5612       1728  ...                     4936   

       stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
count                    8124                   8124      8124       8124   
unique                      9                      9         1          4   
top                         w                      w         p          w   
freq                     4464                   4384      8124       7924   

       ring-number ring-type spore-print-color population habitat  
count         8124      8124              8124       8124    8124  
unique           3         5                 9          6       7  
top              o         p                 w          v       d  
freq          7488      3968              2388       4040    3148  

[4 rows x 23 columns]

In [ ]:
df.dtypes #Check type of each column

class                       object
cap-shape                   object
cap-surface                 object
cap-color                   object
bruises                     object
odor                        object
gill-attachment             object
gill-spacing                object
gill-size                   object
gill-color                  object
stalk-shape                 object
stalk-root                  object
stalk-surface-above-ring    object
stalk-surface-below-ring    object
stalk-color-above-ring      object
stalk-color-below-ring      object
veil-type                   object
veil-color                  object
ring-number                 object
ring-type                   object
spore-print-color           object
population                  object
habitat                     object
dtype: object

In [ ]:
df.isnull().sum() #Checking if empty records exist

class                       0
cap-shape                   0
cap-surface                 0
cap-color                   0
bruises                     0
odor                        0
gill-attachment             0
gill-spacing                0
gill-size                   0
gill-color                  0
stalk-shape                 0
stalk-root                  0
stalk-surface-above-ring    0
stalk-surface-below-ring    0
stalk-color-above-ring      0
stalk-color-below-ring      0
veil-type                   0
veil-color                  0
ring-number                 0
ring-type                   0
spore-print-color           0
population                  0
habitat                     0
dtype: int64

In [ ]:
#some libraries for building graphs
import numpy as np
import plotly.graph_objects as go
from plotly.subplots import make_subplots

#Determine the number of rows and columns for future charts, and also their names
rows = 8
cols = 3
columns = ['class', 'cap-shape', 'cap-surface', 'cap-color', 'bruises', 'odor',
       'gill-attachment', 'gill-spacing', 'gill-size', 'gill-color',
       'stalk-shape', 'stalk-root', 'stalk-surface-above-ring',
       'stalk-surface-below-ring', 'stalk-color-above-ring',
       'stalk-color-below-ring', 'veil-type', 'veil-color', 'ring-number',
       'ring-type', 'spore-print-color', 'population', 'habitat']
fig = make_subplots(rows=rows, cols=cols,subplot_titles=columns ) #Create an area for our charts

#For each property, we build a distribution histogram
for i, col in enumerate(columns):
    r = int(np.ceil(((i+1)/cols)))
    c = i%3+1
    fig.append_trace(go.Histogram(x=df[col]), row=r,col=c)

    fig.update_layout(
    showlegend=False,
    autosize=False,
    width=1300,
    height=1300,
    title={
    'text':'<b>Overview of Distribution of Columns</b>',
         'y':0.95, 
         'x':0.5,
         'xanchor': 'center',
         'yanchor': 'top' 
    })
fig.show()

We see that some features do not occur at all, for example, veil-type is only partial and some other features are also not represented in this dataset. 

Let's see on the distribution of features for target column "class"

In [ ]:
#build another area for our histograms relative to the target class
rows = 8
cols = 3
columns = ['cap-shape', 'cap-surface', 'cap-color', 'bruises', 'odor',
       'gill-attachment', 'gill-spacing', 'gill-size', 'gill-color',
       'stalk-shape', 'stalk-root', 'stalk-surface-above-ring',
       'stalk-surface-below-ring', 'stalk-color-above-ring',
       'stalk-color-below-ring', 'veil-type', 'veil-color', 'ring-number',
       'ring-type', 'spore-print-color', 'population', 'habitat']
fig = make_subplots(rows=rows, cols=cols,subplot_titles=columns )

for i, col in enumerate(columns):
    r = int(np.ceil(((i+1)/cols)))
    c = i%3+1
    p = df[df['class']=='p']
    e = df[df['class']=='e']
    fig.append_trace(go.Histogram(x=p[col],marker_color = '#60391d',name='Poisonous'), row=r,col=c)
    fig.append_trace(go.Histogram(x=e[col],marker_color = '#ff9000',name='Edible'), row=r,col=c)

    
    fig.update_layout(
    showlegend=False,
    autosize=False,
    width=1300,
    height=1700,
    title={
    'text':'<b>Overview of Distribution of Columns for Poisonious & Edible Mushrooms</b>',
         'y':0.98, 
         'x':0.5,
         'xanchor': 'center',
         'yanchor': 'top' 
    })
fig.show()

From here we see that some features absolutely accurately indicate the edible type of mushroom, for example, the absence of odor indicates the edibility of the fungus.

In [ ]:
#create a separate array with the target class and clearly label the classes
#and then remove it from the main dataset and create a separate column for each feature
y = df["class"].map({"p":"poisonous","e":"edible"})
X = df.drop(["class"], axis=1)
X = pd.get_dummies(X, X.columns)
X.head()

cap-shape_b  cap-shape_c  cap-shape_f  cap-shape_k  cap-shape_s  \
0            0            0            0            0            0   
1            0            0            0            0            0   
2            1            0            0            0            0   
3            0            0            0            0            0   
4            0            0            0            0            0   

   cap-shape_x  cap-surface_f  cap-surface_g  cap-surface_s  cap-surface_y  \
0            1              0              0              1              0   
1            1              0              0              1              0   
2            0              0              0              1              0   
3            1              0              0              0              1   
4            1              0              0              1              0   

   ...  population_s  population_v  population_y  habitat_d  habitat_g  \
0  ...             1             0             0          0          0   
1  ...             0             0             0          0          1   
2  ...             0             0             0          0          0   
3  ...             1             0             0          0          0   
4  ...             0             0             0          0          1   

   habitat_l  habitat_m  habitat_p  habitat_u  habitat_w  
0          0          0          0          1          0  
1          0          0          0          0          0  
2          0          1          0          0          0  
3          0          0          0          1          0  
4          0          0          0          0          0  

[5 rows x 117 columns]

In [ ]:
y

0       poisonous
1          edible
2          edible
3       poisonous
4          edible
          ...    
8119       edible
8120       edible
8121       edible
8122    poisonous
8123       edible
Name: class, Length: 8124, dtype: object

In [ ]:
#Using model_selection for splitting the dataset for train and test parts in a ratio of 70 to 30
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1)

The random forest algorithm was chosen as a model, since it is a very universal algorithm, and when adjusting hyperparameters, a high degree of learning can be achieved.

In [ ]:
#import from sklearn library: RandomForestClassifier - our model and GridSearchCV - for automatic selection of hyperparameters for models
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
model = RandomForestClassifier() #create the model
#Setting hyperparameters 
parameters = {
    'n_estimators': [50, 100, 150, 200],  #In our case - the number of trees - the more trees, the better the quality, but the RF setup and operation time also increases proportionally.
    'max_depth': [3, 5, 7, 10, 12, 15],   #Tree depth
    'min_samples_split': [5, 10, 15],     #The minimum number of objects at which splitting is performed
    'min_samples_leaf': [5, 10, 15]       #Limit on the number of objects in leaves
}

rfc = GridSearchCV(model, parameters, refit=True, cv=10)
# Train the model
rfc.fit(X_train, y_train)
# Run the trained model on the test dataset
y_pred = rfc.predict(X_test)

In [ ]:
#for i in range(len(y_pred)):
  #print(y_test.iat[i], y_pred[i])

#import metrics for Evaluating the trained model
from sklearn import metrics
print("Random Forest Classifier model accuracy:", metrics.accuracy_score(y_test, y_pred)*100) #Compare test data with predictions and print an accuracy

Random Forest Classifier model accuracy: 99.79491386382281


Accurancy of the model = 99.79. The model was well trained.

If I had more time, I would probably train a few different models and compare them, and maybe write something as a interface for users to enter their parameters and find out if a given mushroom can be eaten. Or you can go far and create a mobile application that recognizes the proposed parameters of mushrooms from a photo, so when picking mushrooms, the user could take a picture of a mushroom and the neural network would recognize what kind of mushroom it is and whether it can be eaten.